number_theory/numeric.h

```diff
@@ -43,6 +45,54 @@ std::pair<T, T> exgcd(T a, T b) {
   return std::make_pair(xa, ya);
 }
 
+// Accumulates values according to the binary representation of |binary|.
+// For each bit of |binary|, from lower to higher, it calls |update| to
+// accumulate the values. The initial value is |init|.
+template <class T, class U>
+U binary_accumulate(T binary,
+                    const U &init,
+                    std::function<void(bool, U &)> update) {
+  static_assert(
+      std::numeric_limits<T>::is_integer && !std::numeric_limits<U>::is_signed,
+      "binary_accumulate argument |n| must be an unsigned integer");
+
+  T current = binary;
+  U result = init;
+  while (current != 0) {
+    bool bit = (current % 2) != 0;
+    current /= 2;
+    update(bit, result);
+  }
+  return result;
+}
+
+// Computes the value of |base| raised to an integer power |exponent|.
+// The time complexity is O(log |exponent|).
+template <class T, class U>
+T ipow(const T &base, U exponent) {
+  static_assert(std::numeric_limits<U>::is_integer,
+                "ipow exponent must be an integer");
+
+  typedef std::pair<T, T> T2;
+  typedef std::make_unsigned_t<U> unsigned_U;
+
+  auto update = [](bool bit, T2 &state) {
+    auto &[result, power] = state;
+    if (bit)
+      result *= power;
+    power *= power;
+  };
+
+  unsigned_U abs_exp = std::abs(exponent);
+  T2 state = binary_accumulate<unsigned_U, T2>(abs_exp, T2(1, base), update);
+  T result = std::move(state.first);
+
+  if (exponent < 0)
```

Comment:
If `base` is an integer, this will return 0 for negative exponents. I think the user would prefer having a floating point result in that case.

Reply:
Notice that base (type T) might be an arbitrary type. It can be a PrimeField in the future, so `1/result` is meaningful. If the user needs floating point result, he is supposed to provide a floating point type as T.